### PR TITLE
(co)Cartesian structures

### PR DESCRIPTION
**Pull request by eparejatobes**
- [x] categories with products
- [x] categories with sums
- [x] derive (symmetric) monoidal structures
  - [x] implement assocs in terms of (co)products
- [x] syntax: This is mostly implemented. Needs some cleaning and minimal tests.
- [x] use only syntax for categories with (co)products when using tuples and sums
- [ ] aliases for (co)cartesian monoidal structure

### CODE REVIEW

**Comment by eparejatobes** [comment]
@agarciamontoro same as in https://github.com/ohnosequences/stuff/pull/52#issuecomment-337520371

File: src/main/scala/stuff/products.scala [hunk 1]
```diff
@@ -1,6 +1,90 @@
...
+abstract
+class Product {
+
+  type On <: Category
+  val  on  : Category.is[On]
+
+  @infix
+  type ×[X <: On#Objects, Y <: On#Objects] <: On#Objects
+
+  type ∗ <: On#Objects
```
Comment (agarciamontoro) [inline src/main/scala/stuff/products.scala]: I don't know what this ∗ means.

File: src/main/scala/stuff/products.scala [hunk 2]
```diff
@@ -1,6 +1,90 @@
...
+  ]
+  : On#C[A,B] × On#C[C,D] -> On#C[A ⊗ C, B ⊗ D] =
+    λ { fg =>
+      product.both(
+        on.composition( product.left[A,C] and fg.left ) and
+        on.composition(product.right[A,C] and fg.right)
+      )
+    }
+
+  // TODO these two are nice exercises
```
Comment (agarciamontoro) [inline src/main/scala/stuff/products.scala]: What are exactly the exercises? I am not sure I understand how those two methods should behave.

File: src/main/scala/stuff/products.scala [hunk 3]
```diff
@@ -1,6 +1,90 @@
...
+    B <: On#Objects,
+  ]
+  // TODO find a convenient aliasing convention for both ×'s
+  : ohnosequences.stuff.×[On#C[X, A], On#C[X, B]] -> On#C[X, A × B]
+
+  def erase[X <: On#Objects]: On#C[X, ∗]
+
+  def  left[A <: On#Objects, B <: On#Objects]: On#C[A × B, A]
+  def right[A <: On#Objects, B <: On#Objects]: On#C[A × B, B]
+}
```
Comment (agarciamontoro) [inline src/main/scala/stuff/products.scala]: These `left` and `right` methods are just the two projections?

File: src/main/scala/stuff/products.scala (hunk 1 above)
Comment (eparejatobes) [reply src/main/scala/stuff/products.scala]: The terminal object.

File: src/main/scala/stuff/products.scala (hunk 3 above)
Comment (eparejatobes) [reply src/main/scala/stuff/products.scala]: Yes!

File: src/main/scala/stuff/products.scala (hunk 2 above)
Comment (eparejatobes) [reply src/main/scala/stuff/products.scala]: Implementing these two associators in terms of the other methods.

Comment (agarciamontoro) [reply src/main/scala/stuff/products.scala]: Yes, but what should they do? It seems like it's something trivial, but I don't see it 😞

Comment (eparejatobes) [reply src/main/scala/stuff/products.scala]: `(A × B) × C` and `A × (B × C)` are *different*, though isomorphic, types. These two associators (should) represent this isomorphism. You can implement them through repeated use of `left`, `right`, and `both`. Draw it first!

**Comment by eparejatobes** [comment]
I'm leaving this out of 0.4.0.

**Comment by eparejatobes** [comment]
I want to have a convenient arity-free syntax for products and sums. Options:

``` scala
// for the product-induced map
val z: X -> (A × B × C) = 
  f ^ g ^ h
  // alternatives?

val u: (A + B + C) -> X =
  f | g | h
  // alternatives?
```

**Comment by eparejatobes** [comment]
There's already too much in here. I'm going to merge this and work on any deficiencies in separate PRs.

**Pull request merged**